Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
++    public void setTrackViewDistance(int i) {
++        this.lastTrackViewDistance = this.trackViewDistance;
++        this.trackViewDistance = MathHelper.clamp(i, 3, 33);
++        if(i != this.lastTrackViewDistance) {
++            if (this.tracker != null) this.tracker.updatePlayer((EntityPlayer) this);
++        }
++        //set pcm vd to highest vd to respect other processes.
++        if (this.trackViewDistance > ((WorldServer) this.world).getChunkProvider().playerChunkMap.viewDistance)
++            ((WorldServer) this.world).getChunkProvider().playerChunkMap.viewDistance = this.trackViewDistance;
++    }
++    public void setViewDistance(int i) {
++        this.lastViewDistance = this.viewDistance;
++        this.viewDistance = MathHelper.clamp(i, 3, this.trackViewDistance);
++    }
++    //Paper end
++
+     @Override
+     public CraftHumanEntity getBukkitEntity() {
+         return (CraftHumanEntity) super.getBukkitEntity();
+@@ -98,6 +121,12 @@ public abstract class EntityHuman extends EntityLiving {
+ 
+         this.setPositionRotation((double) blockposition.getX() + 0.5D, (double) (blockposition.getY() + 1), (double) blockposition.getZ() + 0.5D, 0.0F, 0.0F);
+         this.aX = 180.0F;
++        //Paper start ticket level pp-view-dist
++        this.setTrackViewDistance(this.world.spigotConfig.viewDistance);
++        this.setViewDistance(this.getTrackViewDistance());
++        this.lastTrackViewDistance = this.getTrackViewDistance();
++        this.lastViewDistance = this.getViewDistance();
++        //Paper end
+     }
+ 
+     public boolean a(World world, BlockPosition blockposition, EnumGamemode enumgamemode) {
+diff --git a/src/main/java/net/minecraft/server/EntityWither.java b/src/main/java/net/minecraft/server/EntityWither.java
+index 8b3052b11..3a00dedfa 100644
+--- a/src/main/java/net/minecraft/server/EntityWither.java
++++ b/src/main/java/net/minecraft/server/EntityWither.java
+@@ -208,9 +208,9 @@ public class EntityWither extends EntityMonster implements IRangedEntity {
+                 // CraftBukkit start - Use relative location for far away sounds
+                 // this.world.b(1023, new BlockPosition(this), 0);
+                 // Paper start
+-                int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
++                //int viewDistance = ((WorldServer) this.world).spigotConfig.viewDistance * 16; // Paper - updated to use worlds actual view distance incase we have to uncomment this due to removal of player view distance API
+                 for (EntityPlayer player : ((WorldServer)world).getPlayers()) {
+-                    //final int viewDistance = player.getViewDistance(); // TODO apply view distance api patch
++                    final int viewDistance = player.getTrackViewDistance(); //Paper ticket level pp-view-dist
```

Comment:
should be in blocks, check the commented line

Reply:
ah noted thanks